Cython/Compiler/ExprNodes.py

```diff
@@ -13564,6 +13594,12 @@ def generate_result_code(self, code):
         code.put_gotref(self.result())
 
 
+class AlreadyAnalysedNode(ExprNode):
+    # Holds some already-analysed subexprs, which are not reanalysed but
+    # are just popped out. Built for "annotations_dict", which is constructed
+    # with its expressions having already been through analyse_types and thus
+    # couldn't be put through again
```

Comment:
While it's not been a strict requirement, we tend to assume in Cython that re-analysing an expression generally works, or at least doesn't break anything. Why is it necessary to prevent it here?

Reply:
The issue was that `CoerceToComplexNode` and `CoerceToBooleanNode` (and possibly some others) didn't have `analyse_types` so the re-analysis failed.

It sounds like it would be better to just give them an `analyse_types` that returns `self` (and now that I look some of the other `CoerceTo...` Nodes have that)? Shall i do that instead?

Reply:
Seems reasonable.